cpp/src/io/csv/writer_impl.hpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file writer_impl.hpp
+ * @brief cuDF-IO CSV writer class implementation header
+ */
+
+#pragma once
+
+#include "csv.h"
+#include "csv_gpu.h"
+
+#include <cudf/strings/strings_column_view.hpp>
+#include <io/utilities/hostdevice_vector.hpp>
+
+#include <cudf/detail/utilities/integer_utils.hpp>
+#include <cudf/io/data_sink.hpp>
+#include <cudf/io/writers.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace cudf {
+namespace experimental {
+namespace io {
+namespace detail {
+namespace csv {
+
+using namespace cudf::io::csv;
```

Comment:
Having `using namespace` in headers tends to be frowned upon since everything in the `cudf::io::csv` and `cudf::io` namespaces will be local to the `cudf::experimental::io::detail::csv` namespace in every TU where this header is included. Maybe it's ok since this is such a deeply nested and specific namespace?

As an alternative, I'd suggest moving the `using namespace` inside the `class writer::impl` declaration.

Reply:
Moving `using namespace` does not compile. See, for example, https://stackoverflow.com/questions/6326805/why-using-namespace-x-is-not-allowed-inside-class-struct-level or https://wandbox.org/permlink/mdMqDZjmVNfDHtkt